wordpress-mobile/WordPress-Android
# Async notifications: media-only error notifications (part III)

Builds on top of #6763 (Post error / success notifications), continuing the implementation of the new design for foreground / error / success notifications for async uploads.

<img width="913" alt="screen shot 2017-10-24 at 4 51 07 pm" src="https://user-images.githubusercontent.com/6597771/31964989-2daaf390-b906-11e7-8098-57ceb860b80e.png">

Notes:
- Please note that, differently to how Posts error notifications are handled (there is one error/final notification per Post, summarising what went well/wrong with a specific Post each time), the notification that covers errors for standalone media items (that is, media that is uploaded through the Media Browser without being included in a Post) is one notification for all.
- Observe that the error notification for media items holds information for all failed standalone (i.e. not-Post related) media items. 
For example, if you are uploading one standalone picture (started uploading through the Media Browser), and it fails, the error notification will show not only the one item that just failed but will tell you all standalone media items that have failed for this site. If you tap RETRY on the notification, you are starting uploads for all of them.
If you prefer to go one by one, you can always pick which to upload from the Media browser, by tapping on the failed items ( “retry” overlay).



To test:

CASE A:
1. start a media upload from the media browser
2. see the progress notification appears and starts progressing
3. turn airplane mode ON
4. see progress notification disappear and error notification come up

![screenshot_20171024-164221](https://user-images.githubusercontent.com/6597771/31965497-deb42a3e-b907-11e7-88e6-72835c2f5140.png)


5. turn airplane mode OFF
6. tap RETRY on the notification / snackbar
7. see the progress notification re-appear and the error notification is gone
8. also see the media browser screen, it should show the same retried item as UPLOADING again.


CASE B:
1. start a media upload from the media browser
2. see the progress notification appears and starts progressing
3. turn airplane mode ON
4. see progress notification disappear and error notification come up
5. turn airplane mode OFF
6. upload yet another file 
7. see both the progress notification for the new file appears, and the error notification for the first file remains.
8. turn airplane mode ON
9. see the progress notification go away, and the error notification is updated now, showing 2 files have not been uploaded.
10. turn airplane mode oFF
11. tap on RETRY
12. see the progress notification re-appear, showing that it’s uploading 2 files,  and the error notification is gone
13. also see the media browser screen, it should show the same retried items (2 items) as UPLOADING again.


CASE C: make a post error appear, and make a stand alone media error appear
1. start a new draft, include some media item, exit the Editor
2. turn airplane mode ON to trigger the error notification
3. turn airplane mode off
4. go to the media browser, and add a new media item
5. when the media item is uploading, turn airplane mode ON 
6. see both error notifications coexist.

<img width="420" alt="screen shot 2017-10-24 at 5 04 10 pm" src="https://user-images.githubusercontent.com/6597771/31965348-64cf9d5c-b907-11e7-867e-2750a92d5bea.png">
_(The image above is the media error notification while the below one is the Post error notification)_

7. Tap on RETRY on the media-only error notification, see the progress starts

<img width="492" alt="screen shot 2017-10-24 at 5 06 15 pm" src="https://user-images.githubusercontent.com/6597771/31965423-aa24ca4e-b907-11e7-911b-eae8a1e07aba.png">


8. Tap on RETRY on the post with media error notification, see the progress info is added to the already existing progress information notification.


CASE D:

1. upload a new media item and turn airplane mode ON in the middle of its upload.
2. turn airplane mode off and try uploading a new item, and again make it fail by turning airplane mode ON in the middle of its upload.
3. you’ll see this second time the error notification is shown, it will read “2 media files not uploaded”, instead of “1 media file not uploaded”..
4. turn airplane mode OFF and tap on RETRY
5. wait until everything is uploaded
6. check the Media Browser and verify your 2 items are uploaded.




cc @nbradbury
